quota/quota_test.go

```diff
@@ -0,0 +1,215 @@
+		if t.err != "" {
+			c.Assert(err, ErrorMatches, t.err, comment)
+			continue
+		}
+		c.Assert(err, IsNil, comment)
+
+		c.Assert(grp.SliceFileName(), Equals, t.name+".slice", comment)
+	}
+}
+
+func (ts *quotaTestSuite) TestSimpleSubGroupVerification(c *C) {
+	tt := []struct {
+		rootlimit quantity.Size
+		subname   string
+		sublimit  quantity.Size
+		err       string
+		comment   string
+	}{
+		{
+			rootlimit: quantity.SizeMiB,
+			subname:   "sub",
+			sublimit:  quantity.SizeMiB,
+			comment:   "basic sub group with same quota as parent happy",
+		},
+		{
+			rootlimit: quantity.SizeMiB,
+			subname:   "sub",
+			sublimit:  quantity.SizeMiB / 2,
+			comment:   "basic sub group with smaller quota than parent happy",
+		},
+		{
+			rootlimit: quantity.SizeMiB,
+			subname:   "sub",
+			sublimit:  quantity.SizeMiB * 2,
+			err:       "sub-group memory limit of 2 MiB is too large to fit inside remaining quota space 1 MiB for parent group myroot",
+			comment:   "sub group with larger quota than parent unhappy",
+		},
+		{
+			rootlimit: quantity.SizeMiB,
+			subname:   "sub-invalid-chars",
+			sublimit:  quantity.SizeMiB,
+			err:       `group name "sub-invalid-chars" contains invalid characters.*`,
+			comment:   "sub group with invalid name",
+		},
+		{
+			rootlimit: quantity.SizeMiB,
+			subname:   "myroot",
+			sublimit:  quantity.SizeMiB,
+			err:       `cannot use same name "myroot" for sub group as parent group`,
+			comment:   "sub group with same name as parent group",
+		},
+		{
+			rootlimit: quantity.SizeMiB,
+			subname:   "snapd",
+			sublimit:  quantity.SizeMiB,
+			err:       `group name "snapd" reserved`,
+			comment:   "sub group with reserved name",
+		},
+		{
+			rootlimit: quantity.SizeMiB,
+			subname:   "zero",
+			sublimit:  0,
+			err:       `group memory limit must be non-zero`,
+			comment:   "sub group with zero memory limit",
+		},
+	}
+
+	for _, t := range tt {
+		comment := Commentf(t.comment)
+		// make a root group
+		rootGrp, err := quota.NewGroup("myroot", t.rootlimit)
+		c.Assert(err, IsNil, comment)
+
+		// make a sub-group under the root group
+		subGrp, err := rootGrp.NewSubGroup(t.subname, t.sublimit)
+		if t.err != "" {
+			c.Assert(err, ErrorMatches, t.err, comment)
+			continue
+		}
+		c.Assert(err, IsNil, comment)
+
+		c.Assert(subGrp.SliceFileName(), Equals, "myroot-"+t.subname+".slice")
+	}
+}
+
+func (ts *quotaTestSuite) TestComplexSubGroups(c *C) {
+	rootGrp, err := quota.NewGroup("myroot", quantity.SizeMiB)
+	c.Assert(err, IsNil)
+
+	// try adding 2 sub-groups with total quota split exactly equally
+	sub1, err := rootGrp.NewSubGroup("sub1", quantity.SizeMiB/2)
+	c.Assert(err, IsNil)
+	c.Assert(sub1.SliceFileName(), Equals, "myroot-sub1.slice")
+
+	sub2, err := rootGrp.NewSubGroup("sub2", quantity.SizeMiB/2)
+	c.Assert(err, IsNil)
+	c.Assert(sub2.SliceFileName(), Equals, "myroot-sub2.slice")
+
+	// adding another sub-group to this group fails
+	_, err = rootGrp.NewSubGroup("sub3", 1)
```

Comment:
Nice! This makes me wonder slightly if in the future we may limit the depth of the nesting but for now I think this is fine